### PR TITLE
Process events in device classes

### PR DESCRIPTION
**Pull request by wmww**
Lot of code churn, but the logic stays roughly the same. This sends the Mir events into `WlKeyboard`/`WlPointer`/`WlTouch` instead of extracting information from them. This is step in the direction of addressing #1792. We want to attach serials to input events, and this moves the events to where we'll need them to do that.

### CODE REVIEW

File: src/server/frontend_wayland/wayland_input_dispatcher.h [hunk 1]
```diff
@@ -66,16 +66,10 @@ class WaylandInputDispatcher
     wayland::Weak<WlSurface> const wl_surface;
 
     std::chrono::nanoseconds timestamp{0};
-    MirPointerButtons last_pointer_buttons{0};
-    std::experimental::optional<std::pair<float, float>> last_pointer_position;
 
     /// Handle user input events
     ///@{
     void handle_keyboard_event(std::chrono::milliseconds const& ms, MirKeyboardEvent const* event);
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wayland_input_dispatcher.h:72]: I think this can go too.

File: src/server/frontend_wayland/wl_keyboard.cpp [hunk 2]
```diff
@@ -20,10 +20,12 @@
 
 #include "wayland_utils.h"
 #include "wl_surface.h"
+#include "wl_seat.h"
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wl_keyboard.cpp:23]: This is just to access `WlSeat::timestamp_of()`?

File: src/server/frontend_wayland/wl_keyboard.cpp [hunk 3]
```diff
@@ -93,20 +112,20 @@ void mf::WlKeyboard::key(std::chrono::milliseconds const& ms, WlSurface* surface
...
             "Sending key 0x%2.2x %s to wl_surface@%u even though it was not explicitly given keyboard focus",
-            scancode, down ? "press" : "release", wl_resource_get_id(surface->resource));
+            scancode, down ? "press" : "release", wl_resource_get_id(surface.resource));
 
         focussed(surface, true);
     }
 }
 
-void mf::WlKeyboard::focussed(WlSurface* surface, bool should_be_focused)
+void mf::WlKeyboard::focussed(WlSurface& surface, bool should_be_focused)
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wl_keyboard.cpp:121]: Why this change of signature? It seems to lead to a lot of needless diff.

File: src/server/frontend_wayland/wl_pointer.cpp [hunk 4]
```diff
@@ -20,6 +20,7 @@
 
 #include "wayland_utils.h"
 #include "wl_surface.h"
+#include "wl_seat.h"
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wl_pointer.cpp:23]: This is just to access WlSeat::timestamp_of()?

File: src/server/frontend_wayland/wl_seat.h [hunk 5]
```diff
@@ -60,6 +63,14 @@ class WlSeat : public wayland::Seat::Global
...
     static auto from(struct wl_resource* seat) -> WlSeat*;
 
+    /// Returns the timestamp of the event in milliseconds, suitable for Wayland events
+    /// @{
+    static auto timestamp_of(MirInputEvent const* event) -> uint32_t;
+    static auto timestamp_of(MirKeyboardEvent const* event) -> uint32_t;
+    static auto timestamp_of(MirPointerEvent const* event) -> uint32_t;
+    static auto timestamp_of(MirTouchEvent const* event) -> uint32_t;
+    /// @}
+
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wl_seat.h:73]: That's a set `static` methods with no obvious connection to `WlSeat`.

1. shouldn't they be at namespace scope?
2. is this the right header? `wl_timestamp.h`?

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
I don't think `WlSeat` is the right place for `timestamp_of()`.

File: src/server/frontend_wayland/wl_seat.h (hunk 5 above)
Comment (wmww) [reply src/server/frontend_wayland/wl_seat.h:73]: Should have explained this better, 2 reasons:
1. In the next phase the seat will manage mapping events to serials, which means it will deal with events and everything that uses events will already be using it
2. The timestamp conversion logic is a little messy and I'd rather not scatter it around the code base else it might get out of sync

I supposed I could make `timestamp_of(MirInputEvent)` on the seat and then `namespace {}` functions for calling it with the various specific types, or I could make `timestamp_of()` a standalone function in `wayland_utils.h` or something.

File: src/server/frontend_wayland/wl_pointer.cpp (hunk 4 above)
Comment (wmww) [reply src/server/frontend_wayland/wl_pointer.cpp:23]: Yes, but it will get included by the next PR regardless (see other comment)

File: src/server/frontend_wayland/wl_keyboard.cpp (hunk 3 above)
Comment (wmww) [reply src/server/frontend_wayland/wl_keyboard.cpp:121]: I guess I didn't need to do that. Did it for consistency with the new code.

File: src/server/frontend_wayland/wl_seat.h (hunk 5 above)
Comment (AlanGriffiths) [reply src/server/frontend_wayland/wl_seat.h:73]: I'm not sure I follow your response.

1. I still don't see why pure functions that take a `Mir*Event` and return a timestamp are members of `WlSeat`, nor why they are declared in the header that defines it. They are perfectly usable without any of `WlSeat`.
2. Why would you think I'm concerned about reducing the number of overloads? They all have the same semantics (I would be worried if they didn't).

Comment (wmww) [reply src/server/frontend_wayland/wl_seat.h:73]: The issue is that I need *somewhere* to put it, though I suppose that isn't the best argument. Is making them standalone functions declared and implemented in the seat files better? Where do you think they should go?

Comment (wmww) [reply src/server/frontend_wayland/wl_seat.h:73]: Perhaps the correct place is the event system.

Comment (AlanGriffiths) [reply src/server/frontend_wayland/wl_seat.h]: OK

**Review by AlanGriffiths — APPROVED** [review]
bors r+
